### ISSUE TITLE
Support for new relic node agent

### ISSUE THREAD

**Issue by theomjones** [nuxt/nuxt.js#6236]
### What problem does this feature solve?

Newrelic's node agent has a requirement that it's the first line of your main module in order to properly bootstrap. https://docs.newrelic.com/docs/agents/nodejs-agent/installation-configuration/install-nodejs-agent#installing

Currently there's no way to do this in Nuxt's setup.

There may also be other use-cases outside of analytics modules, but I can't think of any right now.

### What does the proposed changes look like?

Support some way of requiring modules as the first line of the main js entrypoint, perhaps from a hook or something that will be injected in the build step? I would provide code examples but I'm not that familiar with Nuxt internals yet.

<div align="right"><sub><em>This feature request is available on <a href="https://cmty.app/nuxt">Nuxt</a> community (<a href="https://cmty.app/nuxt/nuxt.js/issues/c9646">#c9646</a>)</em></sub></div>

**Comment by pi0** [comment]
Hi. As the first attempt, did you tried to add `import 'newrelic'` to the top of `nuxt.config.js`? This should partially work as this is almost the top most entrypoint. If it does not work, we may create a custom cli wrapper in the root of project:

`~/nuxt.js`:

```js
require('newrelic')
require('@nuxt/cli')
````

And then inside `package.json`:

```diff
scripts: {
 -- "start": "nuxt start",
 ++ "start": "node ./nuxt start",
}
```

**Comment by pi0** [comment]
Something to add, if newrelic is similar to sentry, this is not **required** to be the first import, we can do this via a nuxt module and it globally should register itself to the entire process.

**Comment by kezek** [comment]
I'm interested in this as well, did you attempt anything @theomjones ?

Thanks

**Comment by theomjones** [comment]
Hi @kezek, yeah we went with the CLI wrapper as new relic has to be the first required module for full metrics.

**Comment by kezek** [comment]
Thanks for the feedback. Do you also use PM2 by any chance ?

**Comment by theomjones** [comment]
Yeah we do, we went by these docs: https://pm2.keymetrics.io/docs/usage/docker-pm2-nodejs/

**Comment by nachogarcia** [comment]
I used the CLI wrapper and it works properly.
Sadly, it's not integrating with the router and all the transactions are shown as /GET
I was thinking of [binding the name in the router](https://docs.newrelic.com/docs/agents/nodejs-agent/api-guides/nodejs-agent-api#request-api) but dind't have time to get my hands into it.

Has anyone faced that issue?

**Comment by nachogarcia** [comment]
After some hassle, I managed to set it up.

I added newrelic as an external webpack dependency in nuxt.config.js
```javascript
    extend (config, ctx) {
      config.externals = [
        { newrelic: 'newrelic' }
      ]
    }
```

And then in nuxtServerInit I set the transaction name with code split so the bundle doesn't break in client side

```javascript
async nuxtServerInit ({ dispatch, commit }, context) {
    const newrelic = await import('newrelic')
    newrelic.setTransactionName(context.route.matched[0].path)
}
```

**Comment by mchagas90** [comment]
@nachogarcia can you see the time cosuming at the external requests on newrelic site? I have the configuration just like yours, but can't see this info.

**Comment by nachogarcia** [comment]
Yeah, they show. Although for some reason I can only see the details of the calls to my PIM, not to my CMS (probably something related to the infra). It looks like this:
![image](https://user-images.githubusercontent.com/9975630/81649244-4c707180-9430-11ea-8587-6d792e3e5881.png)

**Comment by sharonpriel** [comment]
@nachogarcia Hi, we tried your approach with implementing the two rows in 'nuxtServerInit' function to see the full path of Browser tracing names:
const newrelic = await import ('newrelic')
newrelic.SetTransactionName(context.route.matched[0].path)

We found out in our CI/CD process build, the script was broken every time when IIS physical path set to folder when we running the 'npm run build' script.

When we removed the two mentioned rows everything was running as well.

The error we get is:
ERROR in ./node_modules/newrelic/lib/system-info.js
Module not found: Error: Can't resolve 'child_process' in 'node_modules\newrelic\lib'

Can you assists please?
Thank you,
SP

**Comment by nachogarcia** [comment]
No idea. Did you set the external webpack dependency?

Also, I improved this a bit. There were some routes without match so now the line looks like
`newrelic.setTransactionName(context.route.matched[0] ? context.route.matched[0].path : context.route.name)`

And the GETs that didn't get logged were the calls to my healthcheck serverMiddleware so I added the env variable `NEW_RELIC_IGNORING_RULES: '^/health'`

**Comment by sharonpriel** [comment]
1. We did set the external webpack dependency as you suggested. I would like to add that our NuxtJS app is running as a Windows service. We can't find a reason why the build fails to import newrelic in the nuxtServerInit. Do you have any other thoughts on that?

2. Thanks you for sharing your improvements. We'll try it once the above issue will be resolved.

**Comment by nachogarcia** [comment]
Doing a quick search it seems like it's a rather common issue with other libraries. People fix it with

```
module.exports.node = {
  child_process: 'empty'
}
```

Maybe you can go through that route, or you can open an issue in https://github.com/newrelic/node-newrelic

**Comment by sharonpriel** [comment]
We tried child_process: 'false' and it didn't worked so I'm not sure if 'empty' will make a difference but I'll give it a shot.

Much appreciated.
